radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/accounts/BerkeleyAccountTxHistoryStore.java

```diff
@@ -0,0 +1,224 @@
+import com.radixdlt.constraintmachine.RawSubstateBytes;
+import com.radixdlt.constraintmachine.SystemMapKey;
+import com.radixdlt.identifiers.AID;
+import com.radixdlt.identifiers.REAddr;
+import com.radixdlt.store.DatabaseEnvironment;
+import com.radixdlt.store.berkeley.BerkeleyAdditionalStore;
+import com.radixdlt.utils.Longs;
+import com.radixdlt.utils.Pair;
+import com.sleepycat.je.Cursor;
+import com.sleepycat.je.CursorConfig;
+import com.sleepycat.je.Database;
+import com.sleepycat.je.DatabaseConfig;
+import com.sleepycat.je.DatabaseEntry;
+import com.sleepycat.je.Get;
+import com.sleepycat.je.OperationStatus;
+import com.sleepycat.je.Transaction;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import static com.google.common.primitives.UnsignedBytes.lexicographicalComparator;
+import static com.sleepycat.je.OperationStatus.NOTFOUND;
+import static com.sleepycat.je.OperationStatus.SUCCESS;
+
+public class BerkeleyAccountTxHistoryStore implements BerkeleyAdditionalStore {
+	private Database accountTxHistory;
+
+	@Override
+	public void open(DatabaseEnvironment dbEnv) {
+		var env = dbEnv.getEnvironment();
+		this.accountTxHistory = env.openDatabase(null, "radix.account_txn_history", new DatabaseConfig()
+			.setAllowCreate(true)
+			.setTransactional(true)
+			.setKeyPrefixing(true)
+			.setBtreeComparator(lexicographicalComparator())
+		);
+	}
+
+	@Override
+	public void close() {
+		if (accountTxHistory != null) {
+			accountTxHistory.close();
+		}
+	}
+
+	public Stream<Pair<AID, Long>> getTxnIdsAssociatedWithAccount(REAddr addr, Long offset) {
+		var cursor = accountTxHistory.openCursor(null, null);
+		var iterator = new Iterator<Pair<AID, Long>>() {
+			private final DatabaseEntry key;
+			private final DatabaseEntry value = new DatabaseEntry();
+			private OperationStatus status;
+			{
+				if (offset != null) {
+					key = key(addr, offset);
+					status = cursor.get(key, value, Get.SEARCH, null) != null ? SUCCESS : OperationStatus.NOTFOUND;
+				} else {
+					var maybeLast = lastOffset(cursor, value, addr);
+					if (maybeLast.isPresent()) {
+						key = maybeLast.get();
+						status = SUCCESS;
+					} else {
+						key = new DatabaseEntry();
+						status = NOTFOUND;
+					}
+				}
+			}
+
+			@Override
+			public boolean hasNext() {
+				return status == SUCCESS;
+			}
+
+			@Override
+			public Pair<AID, Long> next() {
+				if (status != SUCCESS) {
+					throw new NoSuchElementException();
+				}
+				var nextOffset = Longs.fromByteArray(key.getData(), REAddr.PUB_KEY_BYTES);
+				var nextValue = AID.from(value.getData());
+				var curStatus = cursor.getPrev(key, value, null);
+				status = curStatus == SUCCESS && keyIsREAddr(key, addr) ? SUCCESS : NOTFOUND;
+				return Pair.of(nextValue, nextOffset);
+			}
+		};
```

Comment:
I'd extract the anonymous class declaration to a new descriptive name. The base of my suggestion is the chapter about functions in the clean code book. Here is a summary https://www.fabrizioduroni.it/2019/07/28/clean-code-functions/.

Reply:
I think it's not applicable in this case since this iterator is strongly tied to the IO resource which needs to be closed after use. Trying to separate the functional part and the IO part before knowing the correct abstractions would be overdesign imo.

Reply:
It is more about having small methods that tell what they are supposed to do very quickly from a high level perspective. Breaking it in smaller blocks that can be extracted to methods that describe what they are doing with their names.  It should be a private method. I think it would be even better if java supported nested methods.

```
public Stream<Pair<AID, Long>> getTxnIdsAssociatedWithAccount(REAddr addr, Long offset) {
		var cursor = accountTxHistory.openCursor(null, null);
		var iterator getIterator... (some descriiptive name)
		return Streams.stream(iterator)
			.onClose(cursor::close);
	}
```